x-pack/plugins/security_solution/public/detections/components/rules/rule_preview/index.tsx

```diff
@@ -129,9 +136,20 @@ const RulePreviewComponent: React.FC<RulePreviewProps> = ({
     }
   }, [spaces]);
 
+  // Raw timeframe as a string
   const [startDate, setStartDate] = useState('now-1h');
   const [endDate, setEndDate] = useState('now');
 
+  // Parsed timeframe as a Moment object
+  const [timeframeStart, setTimeframeStart] = useState(moment().subtract(1, 'hour'));
+  const [timeframeEnd, setTimeframeEnd] = useState(moment());
+
+  useEffect(() => {
+    const { start, end } = refreshedTimeframe(startDate, endDate);
+    setTimeframeStart(start);
+    setTimeframeEnd(end);
+  }, [startDate, endDate]);
```

Comment:
Shouldn't `timeframeStart` and `timeframeEnd` be derived from `startDate` and `endDate` using `useMemo`?

Reply:
`startDate` and `endDate` could have a relative time format (like `now`, `now-15m` etc.). Thus, on each preview button click we need to recalculate the real timestamp corresponding to relative time. If we would use `useMemo` the `timeframeStart` and `timeframeEnd` would never change.

Reply:
Okay, I see now, thanks for the explanation 👍